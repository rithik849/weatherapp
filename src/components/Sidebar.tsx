import React, { useState } from 'react'
import {FiMenu, FiArrowLeft} from "react-icons/fi"
import { locationToLatitudeLongitude } from '../constants'
import SidebarItem from './SidebarItem'
import '../styles/Sidebar.css'

const Sidebar : React.FC = () => {
    const [sidebar,setSidebar] = useState(false)

    function toggleSideBar() {
        setSidebar(prev => !prev)
        
    }

    return (
        <div className={sidebar ? "sidebar active" : "sidebar"} >

            {        
                (
                sidebar &&
                <nav className="navigation">
                    {
                    Object.keys(locationToLatitudeLongitude).map(
                        (value,index) => <SidebarItem key={index} country={value}/>
                    )
                    }
                </nav>
                )
            }

            {sidebar ? 
            (<button role="button" className='toggle' onClick={toggleSideBar} >
                <FiArrowLeft className='graphic' />
                 <p className="toggleText">Close Menu</p>
            </button>)
            : 
            (<button role="button" className='toggle' onClick={toggleSideBar}>
                <FiMenu  className='graphic' /> 
                <p className="toggleText">Menu</p>
            </button>)
            }

        
        </div>
    )

}


export default Sidebar