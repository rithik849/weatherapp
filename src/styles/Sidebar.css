
.graphic{
    width: 50px;
    height: 50px;
}

.toggle{
    display: inline;
    align-items: center;
    background-color: black;
    height: fit-content;
}

.toggleText{
    width: auto;
}

.sidebar-item{
    margin-top: 10px;
    margin-bottom: 10px;
}

.selected{
    background-color: blue;
}

.navigation{
    background-color: gray;
    width : 150px;
}