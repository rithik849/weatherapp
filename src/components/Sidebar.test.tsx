import React from 'react';
import renderer from 'react-test-renderer';
import Sidebar from './Sidebar';
import userEvent from '@testing-library/user-event'
import {render, fireEvent, screen, getByText, getByRole} from '@testing-library/react'
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup';

type OmitFirstArg<F> = F extends (x: any, ...args: infer P) => infer R ? (...args: P) => R : never

let fragment : (() => (DocumentFragment | null)) = () => null;
let byText : ( OmitFirstArg<typeof getByText> | (() => null)) = () => null;
let byRole : (OmitFirstArg<typeof getByRole>) | (() => null) = () => null;
let user : UserEvent = userEvent.setup()


describe("Test sidebar interactions", () => {

    beforeEach(() => {
        const {container,asFragment,getByText,getByRole} = render(<Sidebar/>);
        fragment = asFragment
        byText = getByText
        byRole = getByRole
        user = userEvent.setup()
    })
    
    test("render the app",async () => {
        expect(fragment()).toMatchSnapshot()
    })

    test("that the sidebar opens",async () => {
        const menu_button = screen.getByRole("button")
    
    
        await user.click(menu_button as HTMLElement)
    
    
    
        expect(fragment()).toMatchSnapshot()
    })

    test("that the sidebar closes",async () => {
        let menu_button = screen.getByRole("button")
    
    
        await user.click(menu_button as HTMLElement)
        
        menu_button = screen.getByRole("button")
        await user.click(menu_button as HTMLElement)
    
        expect(fragment()).toMatchSnapshot()
    })


})

