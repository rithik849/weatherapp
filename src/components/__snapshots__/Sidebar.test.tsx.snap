// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test sidebar interactions render the app 1`] = `
<DocumentFragment>
  <div
    class="sidebar"
  >
    <button
      class="toggle"
      role="button"
    >
      <svg
        class="graphic"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="3"
          x2="21"
          y1="12"
          y2="12"
        />
        <line
          x1="3"
          x2="21"
          y1="6"
          y2="6"
        />
        <line
          x1="3"
          x2="21"
          y1="18"
          y2="18"
        />
      </svg>
      <p
        class="toggleText"
      >
        Menu
      </p>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Test sidebar interactions that the sidebar closes 1`] = `
<DocumentFragment>
  <div
    class="sidebar"
  >
    <button
      class="toggle"
      role="button"
    >
      <svg
        class="graphic"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="3"
          x2="21"
          y1="12"
          y2="12"
        />
        <line
          x1="3"
          x2="21"
          y1="6"
          y2="6"
        />
        <line
          x1="3"
          x2="21"
          y1="18"
          y2="18"
        />
      </svg>
      <p
        class="toggleText"
      >
        Menu
      </p>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Test sidebar interactions that the sidebar opens 1`] = `
<DocumentFragment>
  <div
    class="sidebar active"
  >
    <nav
      class="navigation"
    >
      <div
        class="sidebar-item "
      >
        Europe/London
      </div>
      <div
        class="sidebar-item "
      >
        Germany
      </div>
    </nav>
    <button
      class="toggle"
      role="button"
    >
      <svg
        class="graphic"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="19"
          x2="5"
          y1="12"
          y2="12"
        />
        <polyline
          points="12 19 5 12 12 5"
        />
      </svg>
      <p
        class="toggleText"
      >
        Close Menu
      </p>
    </button>
  </div>
</DocumentFragment>
`;
